Placement reports

rae-stanton opened the pull request:
This adds the placement report CSV for the current casa organization with tests for the service object as well. This also adds the expected route on the reports page and a decorator to format the date within the CSV to look cleaner. Please let me know if there is anything I may have missed or if something isn't 100% right! 😊😊 

### What github issue is this PR for, if any?
Resolves #4718 
This was part of the new Placements epic! 

### What changed, and why?
This adds the ability to export placement data! Previously, this was not available as an option for the orgs.

### How will this affect user permissions?
- Supervisor/admin permissions: Adds ability to export the data

### How is this tested? (please write tests!) 💖💪
Tested to expect specific columns to render properly when the CSV is generated.


### Screenshots please :)
![placementsample](https://user-images.githubusercontent.com/101673900/233212310-d22903ef-0807-47e2-8808-635555537bd7.jpg)


### Feelings gif~
What gif best describes your feeling working on this issue? 
<img src="https://media2.giphy.com/media/wwg1suUiTbCY8H8vIA/giphy-downsized-medium.gif"/>

rae-stanton committed:
Placement reports

This adds the placement report CSV for the current casa organization with tests for the service object as well. This also adds the expected route on the reports page and a decorator to format the date within the CSV to look cleaner.

rae-stanton committed:
Placement reports

This adds the placement report CSV for the current casa organization with tests for the service object as well. This also adds the expected route on the reports page and a decorator to format the date within the CSV to look cleaner.

compwron commented:
nice gif xD

littleforest commented:
@rae-stanton could you merge in main and fix linting, both standardrb and erb?

rae-stanton committed:
Updates association for placement testing

This updated the casa org association that allows placement service & spec to run as expected.

rae-stanton committed:
lint fix for index erb file

adjusted index file to fix linting issue and also adjusted service file for small rubocop fix.

rae-stanton committed:
lint fix for index erb file

adjusted index file to fix linting issue and also adjusted service file for small rubocop fix.

FireLemons commented:
Try prefixing your commands with `bundle exec` this forces your rails version to match the project's rails version. For example `bundle exec standardrb --fix` If you have a global rails version it will use that by default.

littleforest commented on spec/services/placement_export_csv_service_spec.rb:
```diff
@@ -0,0 +1,16 @@
+  let(:casa_org) { create(:casa_org, name: "Fake Name", display_name: "Fake Display Name") }
+  let(:placement_type) { build(:placement_type, casa_org: casa_org) }
+  let(:creator) { build(:user) }
+  let(:placement) { build(:placement, creator: creator, placement_type: placement_type) }
```
@rae-stanton would you mind just defining all of these within the `it` block instead of using `let`? It makes the specs easier to follow.

littleforest commented on app/views/reports/index.html.erb:
```diff
@@ -63,8 +63,19 @@
+        </li>
       </ul>
     </div>
+  </div>
```
Was there a closing div missing from before? This looks like a bonus div closure.

rae-stanton replied:
I had removed a UL closing tag that wasn't necessary, I think this may be a bonus

littleforest commented on app/controllers/placement_reports_controller.rb:
```diff
@@ -0,0 +1,15 @@
+
+  def index
+    authorize :application, :see_reports_page?
+    placement_report = PlacementExportCsvService.new(casa_org_id: current_organization.id).perform
```
How about we just pass the organization itself instead of the id?
```suggestion
    placement_report = PlacementExportCsvService.new(casa_org: current_organization).perform
```

littleforest commented on app/services/placement_export_csv_service.rb:
```diff
@@ -0,0 +1,39 @@
+
+  def initialize(casa_org_id:)
+    @casa_org = CasaOrg.find(casa_org_id)
+  end
```
Recommend just passing the org itself:

```
def initialize(casa_org)
  @casa_org = casa_org
end
```

littleforest replied:
My bad:

```
def initialize(casa_org:)
  @casa_org = casa_org
end
```

littleforest commented on app/services/placement_export_csv_service.rb:
```diff
@@ -0,0 +1,39 @@
+
+    CSV.generate(headers: true) do |csv|
+      csv << full_data.keys.map(&:to_s).map(&:titleize)
+      if placements.present?
```
You can actually leave off `if placements.present?` since `placements` will be an array, and it just won't loop if it's empty.

littleforest reviewed (commented):
Thanks for working on this!

rae-stanton committed:
Lint fix & updated code changes

This adds the suggested changes from Jeanine - removes `id` in controller to just reference the organization itself and got rid of some unnecessary code in the index and csv service file. Adjusted tests to reflect changes and moved the `let` statements into the `it` block instead.

rae-stanton commented:
Ok all changes have been made! It looks like the only not passing thing with spec checker is that it really wants me to have a controller test. If that's needed, I'm happy to write one up, otherwise, let me know if there's anything else! Thanks for all the help and feedback. @littleforest

littleforest commented:
@rae-stanton it's okay to add the the controller to the list of tests to skip in the `.allow_skipping_tests` file, but would you mind adding a simple request spec, similar to this code: https://github.com/rubyforgood/casa/blob/main/spec/requests/hearing_types_spec.rb#L1-L13

We should fix up that `.allow_skipping_tests` file since we want to get away from controller specs anyway.

littleforest commented on spec/services/placement_export_csv_service_spec.rb:
```diff
@@ -0,0 +1,15 @@
+require "rails_helper"
+require "factory_bot_rails"
```
You can remove this line.

rae-stanton committed:
Test additions

This adds the placement reports controller to the list of skipped controller tests and adds a request spec.

rae-stanton committed:
lint
